### PR TITLE
Update: API method for start_date and stop_date Refs. #1924

### PR DESCRIPTION
**Pull request by APiligrim**
Hello @atodorov, could you please give a recommendation or any advice on how to write test cases for the new start_date and renamed stop_date changes? Also should I add the new property of TestExecution.expected_duration or not yet?

Also for the current commit I made, I was referencing this task from the issue:
`API method TestExecution.update() or the UpdateExecutionForm() - upon changing the status field must set stop_date to the current timestamp if status.weight != 0 (API fix for #1820)`

### CODE REVIEW

File: tcms/rpc/api/testexecution.py [hunk 1]
```diff
@@ -204,6 +204,11 @@ def update(execution_id, values, **kwargs):
     # if this call updated TE.status then adjust timestamps
     if values.get("status"):
         now = timezone.now()
+
+        if form.status.weight != 0:
+            test_execution.stop_date = now
+        test_execution.form.save()
+
```
Comment (atodorov) [inline tcms/rpc/api/testexecution.py:211]: This is already solved on the next few lines. I've updated the issue description so this code block isn't needed.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by atodorov** [comment]
> Hello @atodorov, could you please give a recommendation or any advice on how to write test cases for the new start_date and renamed stop_date changes? 

Do the existing tests in `tcms/rpc/tests/test_testexecution.py` help ? 

> Also should I add the new property of TestExecution.expected_duration or not yet?

This is still blocked on #2253 so you can focus on other parts of this PR.

**Comment by APiligrim** [comment]
Thank you! I checked the `tcms/rpc/tests/test_testexecution.py` and saw that stop_date and start_date are already in TestExecutionFilter() and TestExecutionUpdate(APITestCase) with a few test cases. Do I need to add additional test cases for them?  

I'm thinking to add a test case for the new property `actual_duration`  to test if the format is correct. Since `actual_duration `is a property, not a field, do I still have to add it to a dictionary that looks something like this (and then write possible test cases for it): 
 
```
def test_update_with_single_caserun(self):
        execution = TestExecutionFactory(tested_by=None)

        result = self.rpc_client.TestExecution.update(
            execution.pk,
            {
                "build": self.build.pk,
                "assignee": self.user.pk,
                "sortkey": 90,
                "start_date": "2021-02-25",
                "stop_date": "2021-02-28 12:12:12",
                 "actual_duration": "2021-02-23",
            },
        )
```

**Comment by atodorov** [comment]
> Thank you! I checked the `tcms/rpc/tests/test_testexecution.py` and saw that stop_date and start_date are already in TestExecutionFilter() and TestExecutionUpdate(APITestCase) with a few test cases. Do I need to add additional test cases for them?
> 
> I'm thinking to add a test case for the new property `actual_duration` to test if the format is correct. Since `actual_duration `is a property, not a field, do I still have to add it to a dictionary that looks something like this (and then write possible test cases for it):
> 
> ```
> def test_update_with_single_caserun(self):
>         execution = TestExecutionFactory(tested_by=None)
> 
>         result = self.rpc_client.TestExecution.update(
>             execution.pk,
>             {
>                 "build": self.build.pk,
>                 "assignee": self.user.pk,
>                 "sortkey": 90,
>                 "start_date": "2021-02-25",
>                 "stop_date": "2021-02-28 12:12:12",
>                  "actual_duration": "2021-02-23",
>             },
>         )
> ```


To answer all of these questions you can do the following:

1) Analyze the existing tests and figure out what scenario they cover. Best to describe in words, e.g "a test scenario which will filter test executions and make sure that stop_date is present in the resulting data structure" or  "a scenario which tries updating a pre-existing TE, passing a timestamp for field X in valid/invalid format should pass/fail"

2) The output of 1 will tell you if something is missing or not

3) Then analyze the behavior of the newly added code. We've talked about the 2 timestamp fields above, how about the new property ? What can we do with it, what sort of behavior does the code behind this property exhibit ? Those answers will tell you if you are missing any tests and what they are.

**Comment by APiligrim** [comment]
oops closed this issue by accident from my terminal when trying to send a pull request, going to fix it right now!

**Comment by APiligrim** [comment]
Im also working on adding another test case for TestExecutionUpdate that will check if the property actual_duration, would work after the stop_date or start_date are modified. 

Additionally, should I add a test case in the TestExecutionFilter() to check if its possible to filter by the actual_duration property? Since actual_duration is a property and not in the database, I was not sure how to filter based on property since the  TestExecutionFilter() has mostly fields which connect to the database.

File: tcms/rpc/tests/test_testexecution.py [hunk 2]
```diff
@@ -471,6 +472,7 @@ def test_update_with_single_caserun(self):
         self.assertEqual(result["build"], self.build.pk)
         self.assertIn("status", result)
 
+
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:475]: See flake8 test job.

File: tcms/rpc/tests/test_testexecution.py [hunk 3]
```diff
@@ -622,6 +624,31 @@ def test_update_with_no_perm(self):
...
             )
 
+    def test_create_actual_duration(self):
+        test_execution = TestExecution()
+
+        test_execution.start_date = datetime.strptime("2020-05-05 09:00:00", "%Y-%m-%d %H:%M:%S")
+        test_execution.stop_date = datetime.strptime("2020-05-06 09:00:00", "%Y-%m-%d %H:%M:%S")
+
+        actual_duration_result = test_execution.stop_date - test_execution.start_date
+        self.assertEqual(test_execution.actual_duration,actual_duration_result)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:634]: This test scenario is not related to the API methods thus needs to be moved into tcms/testcases/tests/test_models.py

File: tcms/rpc/tests/test_testexecution.py [hunk 4]
```diff
@@ -622,6 +624,31 @@ def test_update_with_no_perm(self):
                 self.execution_1.pk, {"stop_date": timezone.now()}
             )
 
+    def test_create_actual_duration(self):
+        test_execution = TestExecution()
+
+        test_execution.start_date = datetime.strptime("2020-05-05 09:00:00", "%Y-%m-%d %H:%M:%S")
+        test_execution.stop_date = datetime.strptime("2020-05-06 09:00:00", "%Y-%m-%d %H:%M:%S")
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:631]: You don't need to parse the dates from string, you can work directly with datetime objects in Python.

File: tcms/rpc/tests/test_testexecution.py [hunk 5]
```diff
@@ -622,6 +624,31 @@ def test_update_with_no_perm(self):
...
+            {
+                "start_date": "2021-02-25 12:12:12",
+                "stop_date": "2021-02-28 12:12:12",
+            },
+        )
+
+        execution.refresh_from_db()
+
+        actual_duration_result = result["stop_date"] - result["start_date"]
+        self.assertEqual(test_execution.actual_duration, actual_duration_result)
```
Comment (atodorov) [inline tcms/rpc/tests/test_testexecution.py:651]: What is the reasoning behind this test scenario compared to the previous one ?

**Review by atodorov — CHANGES_REQUESTED** [review]

**Comment by atodorov** [comment]
> Im also working on adding another test case for TestExecutionUpdate that will check if the property actual_duration, would work after the stop_date or start_date are modified.

Before working on anything else I would like to see the results of your analysis of the existing test suite. Can you describe what test scenarios you have found to exist and whether or not you think there is something which is missing ? 

> 
> Additionally, should I add a test case in the TestExecutionFilter() to check if its possible to filter by the actual_duration property? Since actual_duration is a property and not in the database, I was not sure how to filter based on property since the TestExecutionFilter() has mostly fields which connect to the database.

You have the answer right there in your own question!

File: tcms/rpc/tests/test_testexecution.py (hunk 5 above)
Comment (APiligrim) [reply tcms/rpc/tests/test_testexecution.py:651]: At first I was thinking this test can update either the start_date or the stop_date fields again with static data and check if the result from the actual_duration property has also been updated. I know there is already a method which checks when start_date and stop_date are crashing if they’re updated: 
`        self.assertEqual(result["start_date"], execution.start_date)
        self.assertEqual(result["stop_date"], execution.stop_date)`
but in the case that I wrote, I was trying to check if actual_duration property would update when either the stop_date or start_date changed and see its expected vs. actual result.

 But now Im thinking this case is not needed since in the previous test, I also gave static data and checked if the actual_duration property works as expected.

Comment (atodorov) [reply tcms/rpc/tests/test_testexecution.py]: So not only the 2 scenarios are checking exactly the same thing, but also `actual_duration` is a property, aka dynamically calculated field of the Class/object. As long as the mathematics behind it are correct it doesn't matter how often or which of the underlying fields are updated not does it matter how they are updated (API, UI, ORM).

File: tcms/testcases/tests/test_models.py [hunk 6]
```diff
@@ -157,3 +159,18 @@ def test_send_mail_to_case_author(self, send_mail):
...
+class TestActualDurationProperty(BasePlanCase):
+    """Test TestExecution.actual_duration"""
+
+    @classmethod
+    def setUpTestData(cls):
+        super(TestActualDurationProperty, cls).setUpTestData()
+
+        cls.test_execution = TestExecution()
+        cls.test_execution.start_date = timezone.today()
+        cls.test_execution.stop_date = timezone.now()
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:172]: Use `django.utils.timezone.now` + a `timedelta` or hardcode the values of the `datetime` objects like you did before but don't parse them from string.

**Review by atodorov — CHANGES_REQUESTED** [review]
Rebase to latest master and address the rest of the failing test jobs.

File: tcms/testcases/tests/test_models.py [hunk 7]
```diff
@@ -157,3 +159,21 @@ def test_send_mail_to_case_author(self, send_mail):
...
+
+    @classmethod
+    def setUpTestData(cls):
+        super(TestActualDurationProperty, cls).setUpTestData()
+
+        cls.test_execution = TestExecution()
+        cls.test_execution.start_date = timezone.now()
+        cls.test_execution.stop_date = timezone.now() + timezone.timedelta(days=10)
+
+    def test_create_actual_duration(self):
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:175]: Naming of test method should reflect the actual behavior that is being checked to make it obvious what is happening. 

That said you have 3 behaviors:
1) start & stop date both have a timestamp value and we want to check that the calculation gives the expected timedelta
2) start_date is None (see null=True, blank=True in the model definition) aka test execution hasn't been started yet
3) stop_date is None aka test execution hasn't been finished yet

For 2) the expected result is None because you can't have a duration if you don't have a beginning. 

For 3) I am not sure ATM, either None or timezone.now() - start_date (e.g. duration from start up to now!). Pick one of them and hard-code it in a test, we'll adjust it later once we figure out how we want to use this property in the rest of the application.

**Review by atodorov — CHANGES_REQUESTED** [review]

**Review by atodorov — CHANGES_REQUESTED** [review]
Also remove the `.DS_Store` file from the PR. We don't need it.

File: tcms/testcases/tests/test_models.py [hunk 8]
```diff
@@ -157,3 +159,28 @@ def test_send_mail_to_case_author(self, send_mail):
...
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        cls.test_execution = TestExecution()
+        cls.test_execution.start_date = timezone.now()
+        cls.test_execution.stop_date = timezone.now() + timezone.timedelta(days=1, seconds=0, microseconds=0)
+
+    def test_calculation_of_actual_duration(self):
+        self.assertEqual(
+            self.test_execution.actual_duration, timezone.timedelta(days=1, seconds=0, microseconds=2)
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:177]: This is unreliable b/c on different executions in CI it can take more or less than 2 microseconds between assigning the values for start_date and stop_date on llines 172, 173.

If you look at the logs from GitHub CI you will see that the resulting delta is 1 day, 4 microseconds instead of the expected 1day, 2 microseconds. 

If you prefer to use `timezone.now()` then `stop_date = start_date + hard-coded timedelta`.  Alternatively
```
start_date = datetime(<hard-coded year, month, day, hours, etc>)
stop_date = datetime(<another hard-coded value>)
```

File: tcms/testcases/tests/test_models.py [hunk 9]
```diff
@@ -157,3 +159,28 @@ def test_send_mail_to_case_author(self, send_mail):
...
+        )
+
+    def test_actual_duration_empty_start_date(self):
+        empty_start_date_actual_duration = TestExecution(start_date=None)
+        self.assertEqual(empty_start_date_actual_duration.actual_duration, None)
+
+    def test_actual_duration_empty_stop_date(self):
+        empty_stop_date_actual_duration = TestExecution(stop_date=None)
+        self.assertEqual( empty_stop_date_actual_duration.actual_duration, None)
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:186]: Last 2 scenarios look OK. The tests are failing because the implementation doesn't take into account that some/all of the values can be None.  Now you have to update the implementation of this property to account for edge cases which should make these 2 test scenarios pass.

**Review by atodorov — CHANGES_REQUESTED** [review]

File: tcms/testruns/models.py [hunk 10]
```diff
@@ -232,6 +228,14 @@ def _get_absolute_url(self):
         # NOTE: this returns the URL to the TestRun containing this TestExecution!
         return reverse("testruns-get", args=[self.run_id])
 
+    @property
+    def actual_duration(self):
+        if self.stop_date is None or self.start_date is None:
+            return None
+        elif self.start_date is None and self.start_date is None:
+            return None
```
Comment (atodorov) [inline tcms/testruns/models.py:236]: This implementation is overzealous, try figuring out why. 

See the hints reported by CodeClimate and also by pylint (with less verbosity).

File: tcms/testcases/tests/test_models.py [hunk 11]
```diff
@@ -157,3 +159,28 @@ def test_send_mail_to_case_author(self, send_mail):
...
+            self.test_execution.actual_duration, timezone.timedelta(days=1)
+        )
+
+    def test_actual_duration_empty_start_date(self):
+        empty_start_date_actual_duration = TestExecution(start_date=None)
+        self.assertEqual(empty_start_date_actual_duration.actual_duration, None)
+
+    def test_actual_duration_empty_stop_date(self):
+        empty_stop_date_actual_duration = TestExecution(stop_date=None)
+        self.assertEqual( empty_stop_date_actual_duration.actual_duration, None)
```
Comment (atodorov) [inline tcms/testcases/tests/test_models.py:186]: flake8

**Review by atodorov — CHANGES_REQUESTED** [review]
